src/data/finsupp/equiv_dfinsupp.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.dfinsupp
+import data.finsupp.basic
+import algebra.group.pi
+import group_theory.submonoid.operations
+
+/-!
+# Equivalence between `finsupp` and `dfinsupp` constrained to the range of `single i`
+
+## Main definitions
+
+* `finsupp.single_submonoid` - the submonoid corresponding to the range of `finsupp.single`.
+* `finsupp.single_submonoid.of` - the obvious map into `finsupp.single_submonoid`, a `→+`.
+* `finsupp.equiv_dfinsupp_single_submonoid` - the `add_equiv` between a `finsupp` and a `dfinsupp`
+  where each fibre is `finsupp.single_submonoid`.
+
+## Implementation notes
+
+Lean has a very hard time finding some typeclasses in this file - so in many cases we have to
+explicitly build them.
+
+-/
+
+variables {ι : Type*} {M : Type*}
+
+noncomputable theory
+
+namespace finsupp
+
+variables (M)
+
+/-- The submonoid corresponding to the range of `finsupp.single`. -/
+abbreviation single_submonoid [add_monoid M] (i : ι) : add_submonoid (ι →₀ M) :=
+(finsupp.single_add_hom i : M →+ (ι →₀ M)).mrange
+
+/-- The natural construction of elements of `finsupp.single_submonoid`. -/
+abbreviation single_submonoid.of [add_monoid M] (i : ι) : M →+ single_submonoid M i :=
+(finsupp.single_add_hom i : M →+ (ι →₀ M)).mrange_restrict
+
+/-- Typeclass resolution can't find this for some reason, it is needed to define
+`dfinsupp.add_comm_monoid_of_finsupp_single_submonoid`. -/
+instance [add_comm_monoid M] (i : ι) : add_comm_monoid (single_submonoid M i) :=
+by apply_instance
+
+end finsupp
+
+/-- Typeclass resolution can't find this for some reason, it is needed by the `≃+` to state
+`finsupp.equiv_dfinsupp_single_submonoid`. -/
+instance dfinsupp.add_comm_monoid_of_finsupp_single_submonoid [add_comm_monoid M] :
+  add_comm_monoid (Π₀ (i : ι), finsupp.single_submonoid M i) :=
+@dfinsupp.add_comm_monoid ι (λ i, finsupp.single_submonoid M i) _
+
+/-- A `finsupp` is equivalent to a `dfinsupp` where each fibre is constrained to lie in
+`finsupp.single_submonoid M i`. -/
+noncomputable def finsupp.equiv_dfinsupp_single_submonoid [decidable_eq ι] [add_comm_monoid M] :
+  (ι →₀ M) ≃+ (Π₀ i : ι, finsupp.single_submonoid M i) :=
```

Comment:
I would expect the following statement
```suggestion
noncomputable def finsupp.equiv_dfinsupp_single_submonoid [decidable_eq ι] [add_comm_monoid M] :
  (ι →₀ M) ≃+ (Π₀ i : ι, M) :=
```
Is there a reason you picked your version?

Reply:
That's probably a good definition to have somewhere (i'm suspicious we already have it), but its not the one I'm interested in - the idea behind this equivalence is that it equates a polynomial with a direct sum of its monomials.

I suppose it would be possible / sensible to split this equivalence into `(ι →₀ M) ≃+ (Π₀ i : ι, M)` and `(Π₀ i : ι, M) ≃+ (Π₀ i : ι, finsupp.single_submonoid M i)`, but I suspect the proofs would look the same anyway.

Reply:
I'll have a go at doing the split.

Reply:
Ah, the equivalence you describe roughly exists as `finsupp_lequiv_direct_sum`

Reply:
I've added the additive version in #7311